src/array.jl

```diff
@@ -0,0 +1,70 @@
+
+type CLArray{T,N} <: CLObject
+    buffer::Buffer{T}
+    size::NTuple{N,Int}
+    t::Bool
+end
+
+typealias CLMatrix{T} CLArray{T,2}
+typealias CLVector{T} CLArray{T,1}
+
+##  main constructors (overload them if you add new fields)
+
+@compat function CLArray{T,N}(ctx::Context, flags::Tuple{Vararg{Symbol}},
+                      hostarray::AbstractArray{T,N})
+    buf = Buffer(T, ctx, flags, hostbuf=hostarray)
+    sz = size(hostarray)
+    CLArray(buf, sz, false)
+end
+
+copy(A::CLArray) = CLArray(A.buffer, A.size, A.t)
+    
+##  convenient constructors
+
+CLArray{T,N}(ctx::Context, hostarray::AbstractArray{T,N}) = 
+    CLArray(ctx, (:rw, :copy), hostarray)
+
+# TODO: OpenCL may have faster equivalent
+Base.zeros(t::Type, ctx::Context, dims...) = CLArray(ctx, zeros(t, dims...))
```

Comment:
This array needs to determine if it is on the host or the device, and then call an appropriate kernel.